<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>TCR Motif Analysis and Visualization • vdjmatchR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="TCR Motif Analysis and Visualization">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">vdjmatchR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/getting-started.html">Getting Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>TCR Matching</h6></li>
    <li><a class="dropdown-item" href="../articles/batch-matching.html">Batch Matching</a></li>
    <li><a class="dropdown-item" href="../articles/scoring-and-scope.html">Scoring and Search Scope</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Database Tools</h6></li>
    <li><a class="dropdown-item" href="../articles/database-management.html">Database Management</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis &amp; Visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/tcr-motif-analysis.html">TCR Motif Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/tcrdist-analysis.html">TCRdist Distance Analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Integration</h6></li>
    <li><a class="dropdown-item" href="../articles/seurat-vdj-integration.html">Seurat + 10x VDJ Integration</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/furlan-lab/vdjmatchR"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>TCR Motif Analysis and Visualization</h1>
                        <h4 data-toc-skip class="author">vdjmatchR
Authors</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/furlan-lab/vdjmatchR/blob/HEAD/vignettes/tcr-motif-analysis.Rmd" class="external-link"><code>vignettes/tcr-motif-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>tcr-motif-analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<ul>
<li>R (&gt;= 4.0)</li>
<li>Rust (&gt;= 1.70) - <a href="https://www.rust-lang.org/tools/install" class="external-link">Install Rust</a>
</li>
<li>R development tools (Rtools on Windows, Xcode Command Line Tools on
macOS)</li>
</ul>
</div>
<div class="section level3">
<h3 id="from-github-in-r">From Github in R<a class="anchor" aria-label="anchor" href="#from-github-in-r"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"furlan-lab/vdjmatchR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="clone-and-install-from-source-in-shell">Clone and install (from source) in shell<a class="anchor" aria-label="anchor" href="#clone-and-install-from-source-in-shell"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/furlan-lab/vdjmatchR.git</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="bu">cd</span> vdjmatchR</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="ex">R</span> CMD INSTALL .</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-motifstack-from-bioconductor">Install motifStack from Bioconductor<a class="anchor" aria-label="anchor" href="#install-motifstack-from-bioconductor"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"motifStack"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates how to create sequence motif logos from
TCR CDR3 sequences using the motifStack package. Motif logos visualize
conserved amino acid patterns within groups of TCRs, such as those
recognizing the same epitope.</p>
<p><strong>Key applications:</strong></p>
<ul>
<li>Visualize conserved positions in epitope-specific TCRs</li>
<li>Compare motifs across different epitopes</li>
<li>Identify shared patterns in viral-reactive TCR repertoires</li>
<li>Analyze motifs in your matched TCR data</li>
</ul>
</div>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://furlan-lab.github.io/vdjmatchR">vdjmatchR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">motifStack</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: grid</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:motifStack':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as.data.frame</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:data.table':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     first, second</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'IRanges'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:data.table':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     shift</span></span>
<span><span class="co">#&gt; Loading required package: XVector</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biostrings'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:grid':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     pattern</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     strsplit</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-and-prepare-vdjdb-data">Load and prepare VDJdb data<a class="anchor" aria-label="anchor" href="#load-and-prepare-vdjdb-data"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the full VDJdb database</span></span>
<span><span class="va">db_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vdjdb_packaged_path.html">vdjdb_packaged_path</a></span><span class="op">(</span>use_fat_db <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vdjdb_open_file.html">vdjdb_open_file</a></span><span class="op">(</span><span class="va">db_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter for human TRB sequences with high confidence</span></span>
<span><span class="va">db_filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_db.html">filter_db</a></span><span class="op">(</span><span class="va">db</span>,</span>
<span>                         species <span class="op">=</span> <span class="st">"HomoSapiens"</span>,</span>
<span>                         gene <span class="op">=</span> <span class="st">"TRB"</span>,</span>
<span>                         min_vdjdb_score <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert to data.table for exploration</span></span>
<span><span class="va">dt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/db_to_table.html">db_to_table</a></span><span class="op">(</span><span class="va">db_filtered</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="identify-epitopes-with-sufficient-tcr-sequences">Identify epitopes with sufficient TCR sequences<a class="anchor" aria-label="anchor" href="#identify-epitopes-with-sufficient-tcr-sequences"></a>
</h2>
<p>For meaningful motif analysis, we need epitopes with multiple TCR
sequences:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count TCRs per epitope</span></span>
<span><span class="va">epitope_counts</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span>, <span class="va">.N</span>, by <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">antigen_epitope</span>, <span class="va">antigen_species</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">N</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">epitope_counts</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="co">#&gt;          antigen_epitope      antigen_species     N</span></span>
<span><span class="co">#&gt;                   &lt;char&gt;               &lt;char&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1:            GILGFVFTL           InfluenzaA  4645</span></span>
<span><span class="co">#&gt;  2:            NLVPMVATV                  CMV  1348</span></span>
<span><span class="co">#&gt;  3:            GLCTLVAML                  EBV   478</span></span>
<span><span class="co">#&gt;  4:            FLRGRAYGL                  EBV   215</span></span>
<span><span class="co">#&gt;  5:             RAKFKQLL                  EBV   175</span></span>
<span><span class="co">#&gt;  6:           ELAGIGILTV          HomoSapiens   142</span></span>
<span><span class="co">#&gt;  7:            RPIIRPATL           InfluenzaB   119</span></span>
<span><span class="co">#&gt;  8:          HPNGYKSLSTL           InfluenzaB    89</span></span>
<span><span class="co">#&gt;  9:             QMMVKAGL           InfluenzaB    70</span></span>
<span><span class="co">#&gt; 10:           TPRVTGGGAM                  CMV    69</span></span>
<span><span class="co">#&gt; 11:           KLVALGINAV                  HCV    64</span></span>
<span><span class="co">#&gt; 12:            YVLDHLIVV                  EBV    63</span></span>
<span><span class="co">#&gt; 13:      PSDKHIKEYLNKIQN PlasmodiumFalciparum    63</span></span>
<span><span class="co">#&gt; 14:      HIKEYLNKIQNSLST PlasmodiumFalciparum    63</span></span>
<span><span class="co">#&gt; 15:      YLNKIQNSLSTEWSP PlasmodiumFalciparum    63</span></span>
<span><span class="co">#&gt; 16:            RPPIFIRRL                  EBV    39</span></span>
<span><span class="co">#&gt; 17: FRDYVDRFYKTLRAEQASQE                HIV-1    37</span></span>
<span><span class="co">#&gt; 18:            CRVLCCYVL                  CMV    31</span></span>
<span><span class="co">#&gt; 19:          RPHERNGFTVL                  CMV    30</span></span>
<span><span class="co">#&gt; 20:          PQPELPYPQPQ     TriticumAestivum    30</span></span>
<span><span class="co">#&gt;          antigen_epitope      antigen_species     N</span></span>
<span></span>
<span><span class="co"># Filter epitopes with at least 10 unique CDR3 sequences</span></span>
<span><span class="va">epitopes_sufficient</span> <span class="op">&lt;-</span> <span class="va">epitope_counts</span><span class="op">[</span><span class="va">N</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Found %d epitopes with &gt;=10 TCR sequences\n"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">epitopes_sufficient</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found 49 epitopes with &gt;=10 TCR sequences</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="helper-function-create-motif-from-cdr3-sequences">Helper function: Create motif from CDR3 sequences<a class="anchor" aria-label="anchor" href="#helper-function-create-motif-from-cdr3-sequences"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function to create position frequency matrix from aligned CDR3 sequences</span></span>
<span><span class="va">create_tcr_motif</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cdr3_sequences</span>, <span class="va">name</span> <span class="op">=</span> <span class="st">"TCR_motif"</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="co"># Remove any NA or empty sequences</span></span>
<span>  <span class="va">cdr3_sequences</span> <span class="op">&lt;-</span> <span class="va">cdr3_sequences</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">cdr3_sequences</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nzchar</a></span><span class="op">(</span><span class="va">cdr3_sequences</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cdr3_sequences</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Need at least 5 sequences to create a meaningful motif"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># Get sequence lengths</span></span>
<span>  <span class="va">seq_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">cdr3_sequences</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Strategy: align by center or use most common length</span></span>
<span>  <span class="va">median_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">seq_lengths</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Filter sequences close to median length (within 2 amino acids)</span></span>
<span>  <span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">seq_lengths</span> <span class="op">-</span> <span class="va">median_length</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="fl">2</span></span>
<span>  <span class="va">filtered_seqs</span> <span class="op">&lt;-</span> <span class="va">cdr3_sequences</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">filtered_seqs</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/warning.html" class="external-link">warning</a></span><span class="op">(</span><span class="st">"Too few sequences after length filtering, using all sequences"</span><span class="op">)</span></span>
<span>    <span class="va">filtered_seqs</span> <span class="op">&lt;-</span> <span class="va">cdr3_sequences</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># Pad sequences to same length (center alignment)</span></span>
<span>  <span class="va">max_len</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">filtered_seqs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">aligned_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">filtered_seqs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">seq</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">len</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">seq</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">len</span> <span class="op">&lt;</span> <span class="va">max_len</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">pad_total</span> <span class="op">&lt;-</span> <span class="va">max_len</span> <span class="op">-</span> <span class="va">len</span></span>
<span>      <span class="va">pad_left</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">pad_total</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span></span>
<span>      <span class="va">pad_right</span> <span class="op">&lt;-</span> <span class="va">pad_total</span> <span class="op">-</span> <span class="va">pad_left</span></span>
<span>      <span class="va">seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="va">pad_left</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span>,</span>
<span>                   <span class="va">seq</span>,</span>
<span>                   <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="va">pad_right</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">seq</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Create AAStringSet</span></span>
<span>  <span class="va">aa_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-class.html" class="external-link">AAStringSet</a></span><span class="op">(</span><span class="va">aligned_seqs</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Create position frequency matrix with probabilities</span></span>
<span>  <span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/letterFrequency.html" class="external-link">consensusMatrix</a></span><span class="op">(</span><span class="va">aa_set</span>, as.prob <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Remove gap row if present</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="st">"-"</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pfm</span> <span class="op">&lt;-</span> <span class="va">pfm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"-"</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>    <span class="co"># Renormalize after removing gaps</span></span>
<span>    <span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">pfm</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># Ensure columns sum to 1.0 (handle any rounding errors)</span></span>
<span>  <span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">pfm</span>, <span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Replace any NaN with 0</span></span>
<span>  <span class="va">pfm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.nan</a></span><span class="op">(</span><span class="va">pfm</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span>  <span class="co"># Convert to pfm object for motifStack</span></span>
<span>  <span class="va">pfm_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">"pfm"</span>, mat <span class="op">=</span> <span class="va">pfm</span>, name <span class="op">=</span> <span class="va">name</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html" class="external-link">colorset</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="st">"AA"</span>, colorScheme <span class="op">=</span> <span class="st">"chemistry"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">pfm_obj</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Function to create motif for a specific epitope</span></span>
<span><span class="va">create_epitope_motif</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dt</span>, <span class="va">epitope_name</span>, <span class="va">antigen_species</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">antigen_species</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">seqs</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op">==</span> <span class="va">epitope_name</span> <span class="op">&amp;</span> <span class="va">antigen_species</span> <span class="op">==</span> <span class="va">antigen_species</span>, <span class="va">cdr3</span><span class="op">]</span></span>
<span>    <span class="va">motif_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">epitope_name</span>, <span class="st">" ("</span>, <span class="va">antigen_species</span>, <span class="st">")"</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">seqs</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op">==</span> <span class="va">epitope_name</span>, <span class="va">cdr3</span><span class="op">]</span></span>
<span>    <span class="va">motif_name</span> <span class="op">&lt;-</span> <span class="va">epitope_name</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="fu">create_tcr_motif</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">seqs</span><span class="op">)</span>, name <span class="op">=</span> <span class="va">motif_name</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-1-single-epitope-motif-logo">Example 1: Single epitope motif logo<a class="anchor" aria-label="anchor" href="#example-1-single-epitope-motif-logo"></a>
</h2>
<p>Let’s create a motif logo for a well-studied CMV epitope:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># CMV pp65 epitope NLVPMVATV - one of the most studied</span></span>
<span><span class="va">cmv_pp65</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op">==</span> <span class="st">"NLVPMVATV"</span> <span class="op">&amp;</span> <span class="va">antigen_species</span> <span class="op">==</span> <span class="st">"CMV"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Found %d unique CDR3 sequences for CMV NLVPMVATV\n"</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cmv_pp65</span><span class="op">$</span><span class="va">cdr3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found 638 unique CDR3 sequences for CMV NLVPMVATV</span></span>
<span></span>
<span><span class="co"># Create motif</span></span>
<span><span class="va">cmv_motif</span> <span class="op">&lt;-</span> <span class="fu">create_epitope_motif</span><span class="op">(</span><span class="va">dt</span>, <span class="st">"NLVPMVATV"</span>, <span class="st">"CMV"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot motif logo</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifLogo.html" class="external-link">plotMotifLogo</a></span><span class="op">(</span><span class="va">cmv_motif</span>,</span>
<span>              font <span class="op">=</span> <span class="st">"helvetica"</span>,</span>
<span>              ic.scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              motifName <span class="op">=</span> <span class="st">"CMV pp65 NLVPMVATV-specific TCR CDR3 Motif"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: Cairo</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span></code></pre></div>
<p><img src="tcr-motif-analysis_files/figure-html/single-motif-1.png" width="700.0000032"></p>
</div>
<div class="section level2">
<h2 id="example-2-compare-motifs-across-epitopes">Example 2: Compare motifs across epitopes<a class="anchor" aria-label="anchor" href="#example-2-compare-motifs-across-epitopes"></a>
</h2>
<p>Compare TCR motifs for different CMV epitopes:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get top CMV epitopes</span></span>
<span><span class="va">cmv_epitopes</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span><span class="va">antigen_species</span> <span class="op">==</span> <span class="st">"CMV"</span>, <span class="va">.N</span>, by <span class="op">=</span> <span class="va">antigen_epitope</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">N</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create motifs for top CMV epitopes with sufficient sequences</span></span>
<span><span class="va">cmv_motif_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cmv_epitopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">epi</span> <span class="op">&lt;-</span> <span class="va">cmv_epitopes</span><span class="op">$</span><span class="va">antigen_epitope</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="va">n</span> <span class="op">&lt;-</span> <span class="va">cmv_epitopes</span><span class="op">$</span><span class="va">N</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">n</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span></span>
<span>      <span class="va">motif</span> <span class="op">&lt;-</span> <span class="fu">create_epitope_motif</span><span class="op">(</span><span class="va">dt</span>, <span class="va">epi</span>, <span class="st">"CMV"</span><span class="op">)</span></span>
<span>      <span class="va">cmv_motif_list</span><span class="op">[[</span><span class="va">epi</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">motif</span></span>
<span>    <span class="op">}</span>, error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Could not create motif for %s: %s"</span>, <span class="va">epi</span>, <span class="va">e</span><span class="op">$</span><span class="va">message</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Plot motifs side by side</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cmv_motif_list</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifLogoStack.html" class="external-link">plotMotifLogoStack</a></span><span class="op">(</span><span class="va">cmv_motif_list</span>,</span>
<span>                     font <span class="op">=</span> <span class="st">"helvetica"</span>,</span>
<span>                     ic.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span></code></pre></div>
<p><img src="tcr-motif-analysis_files/figure-html/compare-cmv-1.png" width="700.0000032"></p>
</div>
<div class="section level2">
<h2 id="example-3-compare-across-viral-species">Example 3: Compare across viral species<a class="anchor" aria-label="anchor" href="#example-3-compare-across-viral-species"></a>
</h2>
<p>Compare TCR motifs for epitopes from different viruses:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select well-represented epitopes from different viruses</span></span>
<span><span class="va">viral_epitopes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  epitope <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NLVPMVATV"</span>, <span class="st">"GLCTLVAML"</span>, <span class="st">"GILGFVFTL"</span><span class="op">)</span>,</span>
<span>  species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CMV"</span>, <span class="st">"EBV"</span>, <span class="st">"InfluenzaA"</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">viral_motifs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">viral_epitopes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">epi</span> <span class="op">&lt;-</span> <span class="va">viral_epitopes</span><span class="op">$</span><span class="va">epitope</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="va">sp</span> <span class="op">&lt;-</span> <span class="va">viral_epitopes</span><span class="op">$</span><span class="va">species</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="va">n</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op">==</span> <span class="va">epi</span> <span class="op">&amp;</span> <span class="va">antigen_species</span> <span class="op">==</span> <span class="va">sp</span>, <span class="va">.N</span><span class="op">]</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">n</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span></span>
<span>      <span class="va">motif</span> <span class="op">&lt;-</span> <span class="fu">create_epitope_motif</span><span class="op">(</span><span class="va">dt</span>, <span class="va">epi</span>, <span class="va">sp</span><span class="op">)</span></span>
<span>      <span class="va">viral_motifs</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">sp</span>, <span class="va">epi</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">motif</span></span>
<span>    <span class="op">}</span>, error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Could not create motif for %s %s: %s"</span>, <span class="va">sp</span>, <span class="va">epi</span>, <span class="va">e</span><span class="op">$</span><span class="va">message</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Plot comparison</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">viral_motifs</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifLogoStack.html" class="external-link">plotMotifLogoStack</a></span><span class="op">(</span><span class="va">viral_motifs</span>,</span>
<span>                     font <span class="op">=</span> <span class="st">"helvetica"</span>,</span>
<span>                     ic.scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span>
<span><span class="co">#&gt; Warning in checkValidSVG(doc, warn = warn): This picture may not have been</span></span>
<span><span class="co">#&gt; generated by Cairo graphics; errors may result</span></span></code></pre></div>
<p><img src="tcr-motif-analysis_files/figure-html/viral-comparison-1.png" width="700.0000032"></p>
</div>
<div class="section level2">
<h2 id="example-4-analyze-motifs-in-your-matched-data">Example 4: Analyze motifs in your matched data<a class="anchor" aria-label="anchor" href="#example-4-analyze-motifs-in-your-matched-data"></a>
</h2>
<p>After matching your TCRs to VDJdb, analyze motifs in your hits:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assuming you have matched TCRs from Seurat object</span></span>
<span><span class="co"># (see Seurat integration vignette for details)</span></span>
<span></span>
<span><span class="co"># Example: you have a data.frame 'hits' with matched TCRs</span></span>
<span><span class="co"># hits &lt;- match_tcr_many_df(db_filtered, query_cdr3s, query_vs, query_js, ...)</span></span>
<span></span>
<span><span class="co"># Group by epitope and create motifs for abundant hits</span></span>
<span><span class="va">epitope_hits</span> <span class="op">&lt;-</span> <span class="va">hits</span><span class="op">[</span>, <span class="va">.N</span>, by <span class="op">=</span> <span class="va">antigen_epitope</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">N</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create motif for your most common epitope hit</span></span>
<span><span class="va">top_epitope</span> <span class="op">&lt;-</span> <span class="va">epitope_hits</span><span class="op">$</span><span class="va">antigen_epitope</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">my_tcrs</span> <span class="op">&lt;-</span> <span class="va">hits</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op">==</span> <span class="va">top_epitope</span>, <span class="va">query_cdr3</span><span class="op">]</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">my_tcrs</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">my_motif</span> <span class="op">&lt;-</span> <span class="fu">create_tcr_motif</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">my_tcrs</span><span class="op">)</span>,</span>
<span>                               name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"My data:"</span>, <span class="va">top_epitope</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Compare to VDJdb reference for same epitope</span></span>
<span>  <span class="va">ref_motif</span> <span class="op">&lt;-</span> <span class="fu">create_epitope_motif</span><span class="op">(</span><span class="va">dt</span>, <span class="va">top_epitope</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Plot comparison</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/plotMotifLogoStack.html" class="external-link">plotMotifLogoStack</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_motif</span>, <span class="va">ref_motif</span><span class="op">)</span>,</span>
<span>                     font <span class="op">=</span> <span class="st">"helvetica"</span>,</span>
<span>                     ic.scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     motifName <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"Your TCRs vs VDJdb Reference: %s"</span>, <span class="va">top_epitope</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-5-cdr3-length-distribution-by-epitope">Example 5: CDR3 length distribution by epitope<a class="anchor" aria-label="anchor" href="#example-5-cdr3-length-distribution-by-epitope"></a>
</h2>
<p>Before creating motifs, visualize CDR3 length distributions:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add CDR3 length column</span></span>
<span><span class="va">dt</span><span class="op">[</span>, <span class="va">cdr3_length</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">cdr3</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Get top epitopes</span></span>
<span><span class="va">top_epitopes</span> <span class="op">&lt;-</span> <span class="va">dt</span><span class="op">[</span>, <span class="va">.N</span>, by <span class="op">=</span> <span class="va">antigen_epitope</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">N</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="va">antigen_epitope</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot length distribution for top epitopes</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dt</span><span class="op">[</span><span class="va">antigen_epitope</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_epitopes</span><span class="op">]</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cdr3_length</span>, fill <span class="op">=</span> <span class="va">antigen_epitope</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">1</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>, position <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">antigen_epitope</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CDR3 Length Distribution by Epitope"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"CDR3 Length (amino acids)"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="tcr-motif-analysis_files/figure-html/length-dist-1.png" width="700.0000032"></p>
</div>
<div class="section level2">
<h2 id="advanced-information-content-analysis">Advanced: Information content analysis<a class="anchor" aria-label="anchor" href="#advanced-information-content-analysis"></a>
</h2>
<p>Analyze position-specific information content:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function to calculate information content per position</span></span>
<span><span class="va">calculate_ic</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pfm_obj</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Extract probability matrix from pfm object</span></span>
<span>  <span class="va">prob_matrix</span> <span class="op">&lt;-</span> <span class="va">pfm_obj</span><span class="op">@</span><span class="va">mat</span></span>
<span></span>
<span>  <span class="co"># Calculate Shannon entropy per position</span></span>
<span>  <span class="co"># Background frequency for amino acids (uniform)</span></span>
<span>  <span class="va">n_aa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">prob_matrix</span><span class="op">)</span></span>
<span>  <span class="va">max_entropy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">n_aa</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">ic_per_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">prob_matrix</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">col</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Remove zeros to avoid log(0)</span></span>
<span>    <span class="va">col</span> <span class="op">&lt;-</span> <span class="va">col</span><span class="op">[</span><span class="va">col</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span>    <span class="va">entropy</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">col</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">col</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">information</span> <span class="op">&lt;-</span> <span class="va">max_entropy</span> <span class="op">-</span> <span class="va">entropy</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">information</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">ic_per_pos</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Create motif and extract IC for CMV epitope</span></span>
<span><span class="va">cmv_motif</span> <span class="op">&lt;-</span> <span class="fu">create_epitope_motif</span><span class="op">(</span><span class="va">dt</span>, <span class="st">"NLVPMVATV"</span>, <span class="st">"CMV"</span><span class="op">)</span></span>
<span><span class="va">ic_values</span> <span class="op">&lt;-</span> <span class="fu">calculate_ic</span><span class="op">(</span><span class="va">cmv_motif</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot information content per position</span></span>
<span><span class="va">ic_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  position <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ic_values</span><span class="op">)</span>,</span>
<span>  information_content <span class="op">=</span> <span class="va">ic_values</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ic_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">position</span>, y <span class="op">=</span> <span class="va">information_content</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"steelblue"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"steelblue"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Information Content per CDR3 Position - CMV NLVPMVATV"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Position"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Information Content (bits)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `linewidth` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span></code></pre></div>
<p><img src="tcr-motif-analysis_files/figure-html/ic-analysis-1.png" width="700.0000032"></p>
</div>
<div class="section level2">
<h2 id="practical-considerations">Practical considerations<a class="anchor" aria-label="anchor" href="#practical-considerations"></a>
</h2>
<div class="section level3">
<h3 id="minimum-sequence-requirements">Minimum sequence requirements<a class="anchor" aria-label="anchor" href="#minimum-sequence-requirements"></a>
</h3>
<p>For reliable motif analysis:</p>
<ul>
<li>
<strong>Minimum 5 sequences</strong>: Required for basic motif</li>
<li>
<strong>10+ sequences</strong>: Recommended for meaningful
patterns</li>
<li>
<strong>20+ sequences</strong>: Better for robust motif
identification</li>
<li>
<strong>50+ sequences</strong>: Ideal for detailed analysis</li>
</ul>
</div>
<div class="section level3">
<h3 id="handling-variable-length-cdr3s">Handling variable length CDR3s<a class="anchor" aria-label="anchor" href="#handling-variable-length-cdr3s"></a>
</h3>
<p>TCR CDR3 sequences have variable length, which complicates
alignment:</p>
<p><strong>Strategies used in this vignette:</strong></p>
<ol style="list-style-type: decimal">
<li>
<strong>Length filtering</strong>: Keep sequences within ±2 amino
acids of median</li>
<li>
<strong>Center alignment</strong>: Pad sequences to align central
positions</li>
<li>
<strong>Alternative</strong>: Use tools like MUSCLE for multiple
sequence alignment</li>
</ol>
</div>
<div class="section level3">
<h3 id="epitope-selection-tips">Epitope selection tips<a class="anchor" aria-label="anchor" href="#epitope-selection-tips"></a>
</h3>
<p>Best epitopes for motif analysis:</p>
<ul>
<li>High sequence count (&gt;10 unique CDR3s)</li>
<li>Well-characterized (CMV, EBV, Influenza epitopes)</li>
<li>High VDJdb confidence scores (≥2)</li>
<li>Restricted by common HLA alleles</li>
</ul>
</div>
<div class="section level3">
<h3 id="color-schemes">Color schemes<a class="anchor" aria-label="anchor" href="#color-schemes"></a>
</h3>
<p>motifStack supports different color schemes:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Chemistry-based coloring (default for amino acids)</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">"pfm"</span>, mat <span class="op">=</span> <span class="va">matrix</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html" class="external-link">colorset</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="st">"AA"</span>, colorScheme <span class="op">=</span> <span class="st">"chemistry"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Hydrophobicity-based</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">"pfm"</span>, mat <span class="op">=</span> <span class="va">matrix</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html" class="external-link">colorset</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="st">"AA"</span>, colorScheme <span class="op">=</span> <span class="st">"hydrophobicity"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Custom colors</span></span>
<span><span class="va">pfm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="st">"pfm"</span>, mat <span class="op">=</span> <span class="va">matrix</span>, color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/colorset.html" class="external-link">colorset</a></span><span class="op">(</span>alphabet <span class="op">=</span> <span class="st">"AA"</span>, colorScheme <span class="op">=</span> <span class="st">"custom"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="interpretation-guidelines">Interpretation guidelines<a class="anchor" aria-label="anchor" href="#interpretation-guidelines"></a>
</h2>
<p><strong>High information content positions</strong>: Indicate
conserved amino acids important for epitope recognition</p>
<p><strong>Variable positions</strong>: May represent: - Degeneracy in
TCR binding - Sampling bias in VDJdb - True diversity in recognition
modes</p>
<p><strong>Central positions</strong>: Often more conserved as they
typically contact peptide-MHC</p>
<p><strong>Flanking positions</strong>: May show more variation</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.4.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Los_Angeles</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    grid      stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [8] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.5.2        Biostrings_2.74.1    GenomeInfoDb_1.42.3 </span></span>
<span><span class="co">#&gt;  [4] XVector_0.46.0       IRanges_2.40.1       S4Vectors_0.44.0    </span></span>
<span><span class="co">#&gt;  [7] BiocGenerics_0.52.0  motifStack_1.50.0    data.table_1.17.6   </span></span>
<span><span class="co">#&gt; [10] vdjmatchR_0.0.0.9000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.3                   bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.6                 magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;   [5] ade4_1.7-23                 matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;   [7] compiler_4.4.3              RSQLite_2.3.9              </span></span>
<span><span class="co">#&gt;   [9] png_0.1-8                   systemfonts_1.2.2          </span></span>
<span><span class="co">#&gt;  [11] vctrs_0.6.5                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [13] stringr_1.5.1               pwalign_1.2.0              </span></span>
<span><span class="co">#&gt;  [15] pkgconfig_2.0.3             crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [17] fastmap_1.2.0               labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [19] caTools_1.18.3              Rsamtools_2.22.0           </span></span>
<span><span class="co">#&gt;  [21] rmarkdown_2.29              tzdb_0.5.0                 </span></span>
<span><span class="co">#&gt;  [23] UCSC.utils_1.2.0            ragg_1.4.0                 </span></span>
<span><span class="co">#&gt;  [25] DirichletMultinomial_1.48.0 bit_4.6.0                  </span></span>
<span><span class="co">#&gt;  [27] xfun_0.52                   zlibbioc_1.52.0            </span></span>
<span><span class="co">#&gt;  [29] cachem_1.1.0                CNEr_1.42.0                </span></span>
<span><span class="co">#&gt;  [31] jsonlite_2.0.0              blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [33] DelayedArray_0.32.0         BiocParallel_1.40.2        </span></span>
<span><span class="co">#&gt;  [35] jpeg_0.1-11                 parallel_4.4.3             </span></span>
<span><span class="co">#&gt;  [37] R6_2.6.1                    bslib_0.9.0                </span></span>
<span><span class="co">#&gt;  [39] stringi_1.8.7               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [41] rtracklayer_1.66.0          GenomicRanges_1.58.0       </span></span>
<span><span class="co">#&gt;  [43] jquerylib_0.1.4             Rcpp_1.0.14                </span></span>
<span><span class="co">#&gt;  [45] SummarizedExperiment_1.36.0 knitr_1.50                 </span></span>
<span><span class="co">#&gt;  [47] base64enc_0.1-3             R.utils_2.13.0             </span></span>
<span><span class="co">#&gt;  [49] readr_2.1.5                 Matrix_1.7-3               </span></span>
<span><span class="co">#&gt;  [51] tidyselect_1.2.1            dichromat_2.0-0.1          </span></span>
<span><span class="co">#&gt;  [53] abind_1.4-8                 yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [55] codetools_0.2-20            curl_6.4.0                 </span></span>
<span><span class="co">#&gt;  [57] lattice_0.22-7              tibble_3.3.0               </span></span>
<span><span class="co">#&gt;  [59] plyr_1.8.9                  withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [61] KEGGREST_1.46.0             Biobase_2.66.0             </span></span>
<span><span class="co">#&gt;  [63] evaluate_1.0.4              desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [65] pillar_1.10.2               MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt;  [67] generics_0.1.4              grImport2_0.3-3            </span></span>
<span><span class="co">#&gt;  [69] RCurl_1.98-1.17             hms_1.1.3                  </span></span>
<span><span class="co">#&gt;  [71] scales_1.4.0                gtools_3.9.5               </span></span>
<span><span class="co">#&gt;  [73] xtable_1.8-4                glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [75] seqLogo_1.72.0              tools_4.4.3                </span></span>
<span><span class="co">#&gt;  [77] TFMPvalue_0.0.9             BiocIO_1.16.0              </span></span>
<span><span class="co">#&gt;  [79] BSgenome_1.74.0             annotate_1.84.0            </span></span>
<span><span class="co">#&gt;  [81] GenomicAlignments_1.42.0    fs_1.6.6                   </span></span>
<span><span class="co">#&gt;  [83] XML_3.99-0.18               Cairo_1.6-2                </span></span>
<span><span class="co">#&gt;  [85] TFBSTools_1.44.0            poweRlaw_1.0.0             </span></span>
<span><span class="co">#&gt;  [87] colorspace_2.1-1            AnnotationDbi_1.68.0       </span></span>
<span><span class="co">#&gt;  [89] GenomeInfoDbData_1.2.13     restfulr_0.0.15            </span></span>
<span><span class="co">#&gt;  [91] cli_3.6.5                   textshaping_1.0.0          </span></span>
<span><span class="co">#&gt;  [93] S4Arrays_1.6.0              dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [95] gtable_0.3.6                R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [97] sass_0.4.10                 digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [99] SparseArray_1.6.2           rjson_0.2.23               </span></span>
<span><span class="co">#&gt; [101] htmlwidgets_1.6.4           farver_2.1.2               </span></span>
<span><span class="co">#&gt; [103] R.oo_1.27.0                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [105] htmltools_0.5.8.1           pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [107] lifecycle_1.0.4             httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [109] GO.db_3.20.0                bit64_4.6.0-1              </span></span>
<span><span class="co">#&gt; [111] MASS_7.3-65</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by vdjmatch Authors.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
